### PR TITLE
ENG-4310/ENG-4312 implement default registry policy for bundle microservices (+ENG-4326)

### PR DESCRIPTION
**Pull request by user-54dbb57a**

### CODE REVIEW

File: src/main/java/org/entando/kubernetes/model/bundle/installable/PluginInstallable.java [hunk 1]
```diff
@@ -69,6 +70,29 @@ private void overridePlugin(EntandoPlugin plugin) {
...
+
+        boolean pluginNotRecordedInCM = pluginDataRepository.findByBundleIdAndPluginName(
+                meta.getBundleId(), meta.getPluginName()
+        ).isEmpty();
+
+        if (pluginNotRecordedInCM) {
+            boolean pluginLinkedInK8S = kubernetesService.isPluginLinked(meta.getPluginCode());
+
+            if (pluginLinkedInK8S) {
+                var pluginDataEntity = composePluginDataEntity(new PluginDataEntity());
```
Comment (entgigi) [inline src/main/java/org/entando/kubernetes/model/bundle/installable/PluginInstallable.java:91]: we should add logs at info or debug level to know what's happening

File: src/main/java/org/entando/kubernetes/model/bundle/processor/PluginProcessor.java [hunk 2]
```diff
@@ -135,9 +139,11 @@ public List<Installable<PluginDescriptor>> process(BundleReader bundleReader, In
 
             for (String filename : descriptorList) {
                 // parse descriptor
-                PluginDescriptor pluginDescriptor = bundleReader.readDescriptorFile(filename, PluginDescriptor.class);
-                pluginDescriptor.getDockerImage()
-                        .setSha256(craneCommand.getImageDigest(pluginDescriptor.getDockerImage().toString()));
+                PluginDescriptor pluginDescriptor = mkAdjustedPluginDescriptor(
```
Comment (entgigi) [inline src/main/java/org/entando/kubernetes/model/bundle/processor/PluginProcessor.java]: mkAdjusted maybe is it too generic? a more standard name could be normalizePluginDescriptor, what do you think ?

File: src/main/java/org/entando/kubernetes/model/bundle/processor/PluginProcessor.java [hunk 3]
```diff
@@ -174,11 +180,28 @@ public List<Installable<PluginDescriptor>> process(List<EntandoBundleComponentJo
...
-                .setDescriptorMetadata(
-                        new DescriptorMetadata(null, null, null, null, component.getComponentId(), null, null));
+    /**
+     * Reads the plugin descriptor and in case adjust its registry address if not present in the plugin image url.
+     */
+    private static PluginDescriptor mkAdjustedPluginDescriptor(BundleReader bundleReader, String filename,
+            CraneCommand craneCommand) throws IOException {
+        var pluginDescriptor = bundleReader.readDescriptorFile(filename, PluginDescriptor.class);
+
+        String fqImageAddress = getFqImageAddress(pluginDescriptor, bundleReader);
```
Comment (entgigi) [inline src/main/java/org/entando/kubernetes/model/bundle/processor/PluginProcessor.java]: getFqImageAddress > getFullyQualifiedImageAddress to someone else fq could not be easy to understand

**Review by firegloves — APPROVED** [review]
LGTM

File: src/main/java/org/entando/kubernetes/model/bundle/processor/PluginProcessor.java [hunk 4]
```diff
@@ -174,11 +180,28 @@ public List<Installable<PluginDescriptor>> process(List<EntandoBundleComponentJo
...
+        pluginDescriptor.setDockerImage(DockerImage.fromString(fqImageAddress));
```
Comment (entgigi) [inline src/main/java/org/entando/kubernetes/model/bundle/processor/PluginProcessor.java:197]: I think we should log at debug level the calculated FullyQualifiedImage before use it

File: src/main/java/org/entando/kubernetes/service/digitalexchange/BundleUtilities.java [hunk 5]
```diff
@@ -621,4 +625,49 @@ public static String composeBundleResourceRootFolter(BundleReader bundleReader)
...
+            return null;
+        }
+        try {
+            if (imageAddress.startsWith(DOCKER_IMAGE_TRANSPORT_PREFIX)) {
+                return DockerImage.fromString(imageAddress.substring(DOCKER_IMAGE_TRANSPORT_PREFIX.length())).getRegistry();
+            } else {
+                return DockerImage.fromString(imageAddress).getRegistry();
+            }
+        } catch (Exception ex) {
+            return null;
```
Comment (entgigi) [inline src/main/java/org/entando/kubernetes/service/digitalexchange/BundleUtilities.java:644]: we should add a log at error level to sign exception

File: src/main/java/org/entando/kubernetes/service/digitalexchange/BundleUtilities.java [hunk 6]
```diff
@@ -621,4 +625,49 @@ public static String composeBundleResourceRootFolter(BundleReader bundleReader)
...
+    public static String determineComponentFqImageAddress(
+            String componentImageAddress,
+            String bundleImageAddress,
+            String fallback) {
+        //~
+        String componentImageRegistry = extractImageAddressRegistry(componentImageAddress);
+        if (!Strings.isNullOrEmpty(componentImageRegistry)) {
+            return componentImageAddress;
+        }
+
```
Comment (entgigi) [inline src/main/java/org/entando/kubernetes/service/digitalexchange/BundleUtilities.java:660]: in general, here, we should insert log at debug level to understand why the method generated a certain FullyQualifiedImageAddress

File: src/main/java/org/entando/kubernetes/service/digitalexchange/crane/CraneCommand.java [hunk 7]
```diff
@@ -67,6 +76,19 @@ public String getImageDigest(String image) {
...
 
+    private static String readProcessOutput(ProcessHandler processHandler, boolean stderr, String fallback) {
+        String res;
+        try {
+            res = String.join("\n", (stderr)
+                    ? processHandler.getErrorLines()
+                    : processHandler.getOutputLines()
+            );
+        } catch (IOException e) {
+            res = fallback;
```
Comment (entgigi) [inline src/main/java/org/entando/kubernetes/service/digitalexchange/crane/CraneCommand.java:91]: we should log error at least at debug level

**Review by entgigi — CHANGES_REQUESTED** [review]
just some re-name and more logging, as you said we should improve the log readability to make easier the problem determination

File: src/main/java/org/entando/kubernetes/model/bundle/processor/PluginProcessor.java (hunk 3 above)
Comment (user-54dbb57a) [reply src/main/java/org/entando/kubernetes/model/bundle/processor/PluginProcessor.java]: > getFqImageAddress > getFullyQualifiedImageAddress to someone else fq could not be easy to understand

FQ is a very common and well accepted short (FQDN etc..)

File: src/main/java/org/entando/kubernetes/service/digitalexchange/crane/CraneCommand.java (hunk 7 above)
Comment (user-54dbb57a) [reply src/main/java/org/entando/kubernetes/service/digitalexchange/crane/CraneCommand.java:91]: btw this is old code refactore in a function

File: src/main/java/org/entando/kubernetes/model/bundle/installable/PluginInstallable.java (hunk 1 above)
Comment (user-54dbb57a) [reply src/main/java/org/entando/kubernetes/model/bundle/installable/PluginInstallable.java:91]: This condition should not happen, so I'd say warning level

File: src/main/java/org/entando/kubernetes/service/digitalexchange/BundleUtilities.java (hunk 6 above)
Comment (user-54dbb57a) [reply src/main/java/org/entando/kubernetes/service/digitalexchange/BundleUtilities.java:660]: I'm not sure, this function is well covered by tests and this is a pretty normal case
We may add debug code on the callers  if necessary

**Review by entgigi — APPROVED** [review]
LGTM

**Pull request merged**